challengeutils/writeups.py

```diff
@@ -0,0 +1,162 @@
+from . import utils
+from . import permissions
+
+
+def append_writeup_to_main_submission(row, syn):
+    '''
+    Helper function that appends the write up synapse id and archived
+    write up synapse id on the main submission
+
+    Args:
+        row: Dictionary row['team'], row['objectId'], row['archived'],
+             row['entityId']
+        syn: synapse object
+    '''
+    if pd.isnull(row['archived']):
+        print("NO WRITEUP: " + row['team'])
+    else:
+        status = syn.getSubmissionStatus(row['objectId'])
+        add_writeup_dict = {
+            'writeUp': row['entityId'], 'archivedWriteUp': row['archived']}
+
+        add_writeup = to_submission_status_annotations(
+            add_writeup_dict, is_private=False)
+        new_status = utils.update_single_submission_status(
+            status, add_writeup)
+        syn.store(new_status)
+
+
+def attach_writeup(syn, writeup_queueid, submission_queueid):
+    '''
+    Attach the write up to the submission queue
+
+    Args:
+        writeup_queueid:   Write up evaluation queue id
+        submission_queueid: Submission queue id
+    '''
+    writeups = list(utils.evaluation_queue_query(
+        syn,
+        "select team, entityId, archived from evaluation_{} "
+        "where status == 'VALIDATED'".format(writeup_queueid)))
+    submissions = list(utils.evaluation_queue_query(
+        syn,
+        "select objectId, team from evaluation_{} "
+        "where status == 'SCORED'".format(submission_queueid)))
+    writeupsdf = pd.DataFrame(writeups)
+    submissionsdf = pd.DataFrame(submissions)
+    submissions_with_writeupsdf = \
+        submissionsdf.merge(writeupsdf, on="team", how="left")
+
+    submissions_with_writeupsdf.apply(
+        lambda row: append_writeup_to_main_submission(row, syn), axis=1)
+
+
+def validate_project(syn, submission, challenge, public=False, admin=None):
+    """
+    Validate a Project submission; errors found are returned in a dict.
+
+    Args:
+        submission - submission ID
+        challenge - Synapse ID of Challenge wiki
+        public - Project should be public (default: False)
+        admin - (optional) Project should be shared with this username/ID
+    """
+    writeup = syn.getSubmission(submission)
+    errors = []
+
+    type_error = _validate_ent_type(writeup)
+    if type_error:
+        errors.append(type_error)
+
+    contents_error = _validate_project_contents(writeup, challenge)
+    if contents_error:
+        errors.append(contents_error)
+
+    try:
+        if public:
+            errors.extend(_validate_public_permissions(syn, writeup))
+
+        if admin is not None:
+            admin_error = _validate_admin_permissions(syn, writeup, admin)
+            if admin_error:
+                errors.append(admin_error)
+
+    except SynapseHTTPError as e:
+        if e.response.status_code == 403:
+            errors.append(
+                "Submission is private; please update its sharing settings.")
```

Comment:
You could put this into its own function which will make it easier to test.  Something like
```
def check_project_permissions(...):
    errors = []
    try:
        if public:
            errors.extend(_validate_public_permissions(syn, writeup))

        if admin is not None:
            admin_error = _validate_admin_permissions(syn, writeup, admin)
            if admin_error:
                errors.append(admin_error)

    except SynapseHTTPError as e:
        if e.response.status_code == 403:
            errors.append(
                "Submission is private; please update its sharing settings.")
        else:
             raise e
    return errors


def test_check_project_permissions_errorcode():
    """Tests correct status code is caught"""
    mocked_403 = SynapseHTTPError("foo", response=Mock(status_code=403))
    with patch.object(writeups, "_validate_public_permissions") as patch_public,\
         patch.object(writeups, "_validate_admin_permissions",
                      side_effect=mocked_403) as patch_public:
        errors = writeups.check_project_permissions(...)
        assert errors == ["Submission is private; please update its sharing settings."]
```

Should probably actually also raise an error here when its not 403, so it doesn't just silently fail.